// @framer/rmaroun.react-simple-chatbot/code/Examples.tsx
import { Override } from "framer"

export const optionDataExample: Override = () => {
    return {
        steps: [
            {
              id: '1',
              message: 'What number I am thinking?',
              trigger: '2',
            },
            {
              id: '2',
              options: [
                { value: 1, label: 'Number 1', trigger: '4' },
                { value: 2, label: 'Number 2', trigger: '3' },
                { value: 3, label: 'Number 3', trigger: '3' },
              ],
            },
            {
              id: '3',
              message: 'Wrong answer, try again.',
              trigger: '2',
            },
            {
              id: '4',
              message: 'Awesome! You are a telepath!',
              end: true,
            },
          ]
    };
}

export const validatorDataExample: Override = () => {
    return {
        steps: [
            {
              id: '1',
              message: 'Please type a number',
              trigger: '2',
            },
            {
              id: '2',
              user: true,
              validator: (value) => {
                if (isNaN(value)) {
                  return 'value should be a number';
                }
                return true;
              },
              trigger: '1',
            },
          ]
    };
}

export const previousValueDataExample: Override = () => {
    return {
        steps: [
            {
              id: '1',
              message: 'What is your name?',
              trigger: '2',
            },
            {
              id: '2',
              user: true,
              trigger: '3',
            },
            {
              id: '3',
              message: 'Hi {previousValue}, nice to meet you!',
              end: true,
            },
          ]
    };
}

export const customComponentDataExample: Override = () => {
    return {
        steps: [
            {
              id: '1',
              message: 'You can add custom components',
              trigger: '2',
            },
            {
              id: '2',
              component: (<div> This is a example component </div>),
              end: true,
            },
          ]
    };
}
